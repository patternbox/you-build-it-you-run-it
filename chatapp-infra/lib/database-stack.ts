// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0

import { RemovalPolicy, Stack, StackProps } from 'aws-cdk-lib'
import { AttributeType, BillingMode, Table, TableEncryption } from 'aws-cdk-lib/aws-dynamodb';
import { Construct } from 'constructs';

export class DatabaseStack extends Stack {

  readonly messagesTable: Table;
  readonly channelsTable: Table;
  readonly connectionsTable: Table;

  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    const sharedDynamoSettings = {
      billingMode: BillingMode.PROVISIONED,
      readCapacity: 1,
      writeCapacity: 1,
      removalPolicy: RemovalPolicy.DESTROY, // NOT recommended for production use
      encryption: TableEncryption.AWS_MANAGED,
      pointInTimeRecovery: false, // set to "true" to enable PITR      
    }

    this.connectionsTable = new Table(this, 'Connections', {
      partitionKey: { name: 'connectionId', type: AttributeType.STRING },
      ...sharedDynamoSettings,
    });

    this.channelsTable = new Table(this, 'serverless-chat-channels', {
      partitionKey: {
        name: 'id',
        type: AttributeType.STRING
      },
      tableName: 'serverless-chat-channels',
      ...sharedDynamoSettings,
    });

    this.messagesTable = new Table(this, 'serverless-chat-messages', {
      partitionKey: {
        name: 'channelId',
        type: AttributeType.STRING
      },
      sortKey: {
        name: 'sentAt',
        type: AttributeType.STRING
      },
      tableName: 'serverless-chat-messages',
      ...sharedDynamoSettings,
    });
  }
};
