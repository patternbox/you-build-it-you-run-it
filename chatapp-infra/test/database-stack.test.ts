import {App} from 'aws-cdk-lib'
import {Template} from 'aws-cdk-lib/assertions'
import {DatabaseStack} from '../lib/database-stack'

const databaseStackProps = {
    env: {
        account: '12345789',
        region: 'aws-region'
    }
}

describe('DatabaseStackTest', () => {

    it('should match snapshot', function () {
        // given
        const app = new App()

        // when
        const testStack = new DatabaseStack(app, 'DatabaseStack', databaseStackProps)

        // then
        const template = Template.fromStack(testStack)
        expect(template).toMatchSnapshot()
    })
})
