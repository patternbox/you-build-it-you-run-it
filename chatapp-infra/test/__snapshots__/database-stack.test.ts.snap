// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatabaseStackTest should match snapshot 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "Connections827CEF2F": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "AttributeDefinitions": Array [
          Object {
            "AttributeName": "connectionId",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "KeySchema": Array [
          Object {
            "AttributeName": "connectionId",
            "KeyType": "HASH",
          },
        ],
        "PointInTimeRecoverySpecification": Object {
          "PointInTimeRecoveryEnabled": false,
        },
        "SSESpecification": Object {
          "SSEEnabled": true,
        },
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Delete",
    },
    "serverlesschatchannels532BEFF0": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "AttributeDefinitions": Array [
          Object {
            "AttributeName": "id",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "KeySchema": Array [
          Object {
            "AttributeName": "id",
            "KeyType": "HASH",
          },
        ],
        "PointInTimeRecoverySpecification": Object {
          "PointInTimeRecoveryEnabled": false,
        },
        "SSESpecification": Object {
          "SSEEnabled": true,
        },
        "TableName": "serverless-chat-channels",
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Delete",
    },
    "serverlesschatmessagesE16DD81A": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "AttributeDefinitions": Array [
          Object {
            "AttributeName": "channelId",
            "AttributeType": "S",
          },
          Object {
            "AttributeName": "sentAt",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "KeySchema": Array [
          Object {
            "AttributeName": "channelId",
            "KeyType": "HASH",
          },
          Object {
            "AttributeName": "sentAt",
            "KeyType": "RANGE",
          },
        ],
        "PointInTimeRecoverySpecification": Object {
          "PointInTimeRecoveryEnabled": false,
        },
        "SSESpecification": Object {
          "SSEEnabled": true,
        },
        "TableName": "serverless-chat-messages",
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Delete",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
